File: logdog/src/main/java/cn/thens/logdog/PrettyLogger.java
package cn.thens.logdog;

import android.util.Log;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * @author 7hens
 */
public class PrettyLogger implements Logger<Object> {
    private static final int CHUNK_SIZE = 1024;
    private final Logger<String> lineLogger;

    public PrettyLogger(Logger<String> lineLogger) {
        this.lineLogger = lineLogger;
    }

    @Override
    public void log(int priority, String tag, Object message) {
        Style style = getStyle(priority, tag);
        lineLogger.log(priority, tag, style.top);
        {
            StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
            int methodOffset = getMethodOffset(priority, tag);
            int methodCount = getMethodCount(priority, tag);
            int stackOffset = getStackOffset(stackTrace) + methodOffset;
            int headerLineCount = Math.min(methodCount, stackTrace.length - 1 - stackOffset);
            for (int i = 0; i < headerLineCount; i++) {
                int stackIndex = i + stackOffset;
                String stackInfo = getStackInfo(stackTrace[stackIndex], i == 0);
                String indent = i == 0 ? "" : "  ";
                lineLogger.log(priority, tag, style.middle + indent + stackInfo);
            }
            if (headerLineCount > 0) {
                lineLogger.log(priority, tag, style.divider);
            }
        }
        {
            String messageText = getMessageText(message);
            if (messageText == null) messageText = "null";
            String[] lines = messageText.split("\n");
            for (int lineIndex = 0; lineIndex < lines.length; lineIndex++) {
                String line = lines[lineIndex];
                if (line.isEmpty()) {
                    if (lineIndex != lines.length - 1) {
                        lineLogger.log(priority, tag, style.middle);
                    }
                    break;
                }
                int lineLength = line.length();
                int chunkSize = CHUNK_SIZE;
                int pageSize = (lineLength - 1) / chunkSize + 1;
                for (int pageIndex = 0; pageIndex < pageSize; pageIndex++) {
                    int offset = pageIndex * chunkSize;
                    int count = Math.min(lineLength - offset, chunkSize);
                    String chunk = line.substring(offset, offset + count);
                    lineLogger.log(priority, tag, pageIndex == 0 ? style.middle + chunk : chunk);
                }
            }
        }
        lineLogger.log(priority, tag, style.bottom);
    }

    protected Style getStyle(int priority, String tag) {
        return priority > Log.DEBUG ? Style.DOUBLE : Style.SINGLE;
    }

    protected int getMethodCount(int priority, String tag) {
        return 1;
    }

    protected int getMethodOffset(int priority, String tag) {
        return 0;
    }

    protected String getMessageText(Object message) {
        return LogMessages.of(message);
    }

    private String getStackInfo(StackTraceElement element, boolean putsThreadInfo) {
        String className = element.getClassName();
        className = className.substring(className.lastIndexOf(".") + 1);

        StringBuilder result = new StringBuilder()
                .append(className).append(".").append(element.getMethodName())
                .append("(").append(element.getFileName()).append(":")
                .append(element.getLineNumber()).append(")");
        if (putsThreadInfo) {
            result.append(" on thread: ").append(Thread.currentThread().getName());
        }
        return result.toString();
    }

    private int getStackOffset(StackTraceElement[] stackTrace) {
        boolean isInOffsetClass = false;
        for (int i = 0; i < stackTrace.length; i++) {
            StackTraceElement traceElement = stackTrace[i];
            try {
                Class<?> cls = Class.forName(traceElement.getClassName());
                if (cls.isAnnotationPresent(Ignored.class)) {
                    isInOffsetClass = true;
                } else if (isInOffsetClass) {
                    return i;
                }
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
        }
        return 0;
    }

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    public @interface Ignored {
    }

    public static class Style {
        private static final String EMPTY = "";
        private static final String SINGLE_PART = "────────────────────────────";
        private static final String DASHED_PART = "┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄";
        private static final String DOUBLE_PART = "════════════════════════════";
        private static final String SINGLE_LINE = SINGLE_PART + SINGLE_PART + SINGLE_PART + SINGLE_PART;
        private static final String DASHED_LINE = DASHED_PART + DASHED_PART + DASHED_PART + DASHED_PART;
        private static final String DOUBLE_LINE = DOUBLE_PART + DOUBLE_PART + DOUBLE_PART + DOUBLE_PART;
        private static final String SINGLE_TOP = "┌" + SINGLE_LINE;
        private static final String SINGLE_DIV = "├" + DASHED_LINE;
        private static final String SINGLE_MID = "│ ";
        private static final String SINGLE_BOT = "└" + SINGLE_LINE;
        private static final String DOUBLE_TOP = "╔" + DOUBLE_LINE;
        private static final String DOUBLE_DIV = "╟" + DASHED_LINE;
        private static final String DOUBLE_MID = "║ ";
        private static final String DOUBLE_BOT = "╚" + DOUBLE_LINE;
        public static final Style SINGLE = new Style(SINGLE_TOP, SINGLE_DIV, SINGLE_MID, SINGLE_BOT);
        public static final Style DOUBLE = new Style(DOUBLE_TOP, DOUBLE_DIV, DOUBLE_MID, DOUBLE_BOT);
        public static final Style NONE = new Style(EMPTY, EMPTY, EMPTY, EMPTY);

        final String top;
        final String divider;
        final String middle;
        final String bottom;

        public Style(String top, String divider, String middle, String bottom) {
            this.top = top;
            this.divider = divider;
            this.middle = middle;
            this.bottom = bottom;
        }
    }
}

File: src/main/java/cn/thens/logdog/PrettyLogger.kt
package cn.thens.logdog

import android.util.Log

/**
 * @author 7hens
 */
class PrettyLogger(
    private val logger: Logger,
    private val adapter: StyleAdapter = StyleAdapter.DEFAULT
) : Logger {
    override fun log(priority: LogPriority, tag: String, message: Any?) {
        val (borders, stackOffset, stackCount) = adapter.getStyle(priority, tag)
        if (borders.top.isNotEmpty()) {
            logger.log(priority, tag, borders.top)
        }
        if (stackCount > 0) {
            val stackTrace = Thread.currentThread().stackTrace
            val stackTraceOffset = getStackTraceOffset(stackTrace) + stackOffset
            val headerLineCount = stackCount.coerceAtMost(stackTrace.size - 1 - stackTraceOffset)
            for (i in 0 until headerLineCount) {
                val stackIndex = i + stackTraceOffset
                var stackInfo = getStackInfo(stackTrace[stackIndex])
                if (i == 0) stackInfo += " on thread: " + Thread.currentThread().name
                val indent = if (i == 0) "" else "  "
                logger.log(priority, tag, borders.middle + indent + stackInfo)
            }
            if (headerLineCount > 0) {
                logger.log(priority, tag, borders.divider)
            }
        }
        val messageText = stringOf(message)
        val lines = messageText.split("\n").toTypedArray()
        for (lineIndex in lines.indices) {
            val line = lines[lineIndex]
            if (line.isEmpty()) {
                if (lineIndex != lines.size - 1) {
                    logger.log(priority, tag, borders.middle)
                }
                continue
            }
            val lineLength = line.length
            val chunkSize = LINE_CHUNK_SIZE
            val pageSize = (lineLength - 1) / chunkSize + 1
            for (pageIndex in 0 until pageSize) {
                val offset = pageIndex * chunkSize
                val count = (lineLength - offset).coerceAtMost(chunkSize)
                val chunk = line.substring(offset, offset + count)
                logger.log(priority, tag, if (pageIndex == 0) borders.middle + chunk else chunk)
            }
        }
        if (borders.bottom.isNotEmpty()) {
            logger.log(priority, tag, borders.bottom)
        }
    }

    data class Style(
        val borders: Borders,
        val stackOffset: Int = 0,
        val stackCount: Int = 1
    )

    fun interface StyleAdapter {
        fun getStyle(priority: LogPriority, tag: String): Style

        companion object {
            val DEFAULT = StyleAdapter { p, _ ->
                Style(if (p >= LogPriority.WARN) Borders.DOUBLE else Borders.SINGLE)
            }
        }
    }

    interface Ignored

    class Borders(val top: String, val divider: String, val middle: String, val bottom: String) {
        companion object {
            private const val EMPTY = ""
            private const val SINGLE_PART = "────────────────────────────"
            private const val DASHED_PART = "┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄"
            private const val DOUBLE_PART = "════════════════════════════"
            private const val SINGLE_LINE = SINGLE_PART + SINGLE_PART + SINGLE_PART + SINGLE_PART
            private const val DASHED_LINE = DASHED_PART + DASHED_PART + DASHED_PART + DASHED_PART
            private const val DOUBLE_LINE = DOUBLE_PART + DOUBLE_PART + DOUBLE_PART + DOUBLE_PART
            private const val SINGLE_TOP = "┌$SINGLE_LINE"
            private const val SINGLE_DIV = "├$DASHED_LINE"
            private const val SINGLE_MID = "│ "
            private const val SINGLE_BOT = "└$SINGLE_LINE"
            private const val DOUBLE_TOP = "╔$DOUBLE_LINE"
            private const val DOUBLE_DIV = "╟$DASHED_LINE"
            private const val DOUBLE_MID = "║ "
            private const val DOUBLE_BOT = "╚$DOUBLE_LINE"

            val SINGLE = Borders(SINGLE_TOP, SINGLE_DIV, SINGLE_MID, SINGLE_BOT)
            val DOUBLE = Borders(DOUBLE_TOP, DOUBLE_DIV, DOUBLE_MID, DOUBLE_BOT)
            val NONE = Borders(EMPTY, EMPTY, EMPTY, EMPTY)
        }
    }

    companion object {
        private const val LINE_CHUNK_SIZE = 1024

        fun getStackTraceOffset(stackTrace: Array<StackTraceElement>): Int {
            var isInOffsetClass = false
            for (i in stackTrace.indices) {
                val traceElement = stackTrace[i]
                try {
                    val cls = Class.forName(traceElement.className)
                    if (cls.isSynthetic) {
                        continue
                    }
                    if (Ignored::class.java.isAssignableFrom(cls)) {
                        isInOffsetClass = true
                    } else if (isInOffsetClass) {
                        return i
                    }
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
            }
            return 0
        }

        fun getStackInfo(element: StackTraceElement): String {
            var className = element.className
            className = className.substring(className.lastIndexOf(".") + 1)
            return className + "." + element.methodName +
                    "(" + element.fileName + ":" +
                    element.lineNumber + ")"
        }

        fun getStackInfo(offset: Int = 0): String {
            val stackTrace = Thread.currentThread().stackTrace
            val stackTraceOffset = getStackTraceOffset(stackTrace) + offset
            return getStackInfo(stackTrace[stackTraceOffset])
        }

        fun stringOf(obj: Any?): String {
            return when (obj) {
                null -> "null"
                is String -> obj
                is Throwable -> Log.getStackTraceString(obj)
                is BooleanArray -> obj.contentToString()
                is ByteArray -> obj.contentToString()
                is CharArray -> obj.contentToString()
                is ShortArray -> obj.contentToString()
                is IntArray -> obj.contentToString()
                is LongArray -> obj.contentToString()
                is FloatArray -> obj.contentToString()
                is DoubleArray -> obj.contentToString()
                is Array<*> -> obj.contentDeepToString()
                else -> obj.toString()
            }
        }
    }
}